From eea2056238a156583968f66a12148f829fb332ca Mon Sep 17 00:00:00 2001
From: Lawrence Rust <lvr@softsystem.co.uk>
Date: Mon, 29 Feb 2016 17:10:06 +0000
Subject: [PATCH 309/348] MythMusic: Make initial scan by CD ripper run in
 background

This avoids haveing the UI block while the CD is being scanned and
a CDDB lookup is made.

Signed-off-by: Lawrence Rust <lvr@softsystem.co.uk>
---
 mythplugins/mythmusic/mythmusic/cdrip.cpp |  129 ++++++++++++++---------------
 mythplugins/mythmusic/mythmusic/cdrip.h   |    1 +
 2 files changed, 61 insertions(+), 69 deletions(-)

diff --git a/mythplugins/mythmusic/mythmusic/cdrip.cpp b/mythplugins/mythmusic/mythmusic/cdrip.cpp
index 639a9cb..3ce1b5a 100644
--- a/mythplugins/mythmusic/mythmusic/cdrip.cpp
+++ b/mythplugins/mythmusic/mythmusic/cdrip.cpp
@@ -560,6 +560,7 @@ Ripper::Ripper(MythScreenStack *parent, QString device) :
 
     m_tracks(new QVector<RipTrack*>),
 
+    m_isCompilation(false),
     m_somethingwasripped(false),
     m_mediaMonitorActive(false),
 
@@ -685,7 +686,7 @@ bool Ripper::Create(void)
     m_qualityList->SetValueByData(qVariantFromValue(
                         gCoreContext->GetNumSetting("DefaultRipQuality", 1)));
 
-    QTimer::singleShot(500, this, SLOT(startScanCD()));
+    QTimer::singleShot(50, this, SLOT(startScanCD()));
 
     return true;
 }
@@ -803,6 +804,12 @@ void Ripper::startScanCD(void)
     QString message = tr("Scanning CD. Please Wait ...");
     OpenBusyPopup(message);
 
+    m_artistName.clear();
+    m_albumName.clear();
+    m_genreName.clear();
+    m_year.clear();
+    m_isCompilation = false;
+
     m_scanThread = new CDScannerThread(this);
     connect(m_scanThread->qthread(), SIGNAL(finished()), SLOT(ScanFinished()));
     m_scanThread->start();
@@ -814,70 +821,16 @@ void Ripper::ScanFinished()
     delete m_scanThread;
     m_scanThread = nullptr;
 
-    m_tracks->clear();
-
-    if (m_decoder)
-    {
-        MusicMetadata *metadata;
-        bool isCompilation = false;
-
-        m_artistName.clear();
-        m_albumName.clear();
-        m_genreName.clear();
-        m_year.clear();
-
-        for (int trackno = 0; trackno < m_decoder->getNumTracks(); trackno++)
-        {
-            RipTrack *ripTrack = new RipTrack;
-
-            metadata = m_decoder->getMetadata(trackno + 1);
-            if (metadata)
-            {
-                ripTrack->metadata = metadata;
-                ripTrack->length = metadata->Length();
-
-                if (metadata->Compilation())
-                {
-                    isCompilation = true;
-                    m_artistName = metadata->CompilationArtist();
-                }
-                else if (m_artistName.isEmpty())
-                {
-                    m_artistName = metadata->Artist();
-                }
-
-                if (m_albumName.isEmpty())
-                    m_albumName = metadata->Album();
-
-                if (m_genreName.isEmpty() && !metadata->Genre().isEmpty())
-                    m_genreName = metadata->Genre();
-
-                if (m_year.isEmpty() && metadata->Year() > 0)
-                    m_year = QString::number(metadata->Year());
-
-                QString title = metadata->Title();
-                ripTrack->isNew = isNewTune(m_artistName, m_albumName, title);
-
-                ripTrack->active = ripTrack->isNew;
+    m_artistEdit->SetText(m_artistName);
+    m_albumEdit->SetText(m_albumName);
+    m_genreEdit->SetText(m_genreName);
+    m_yearEdit->SetText(m_year);
+    m_compilationCheck->SetCheckState(m_isCompilation);
 
-                m_tracks->push_back(ripTrack);
-
-            }
-            else
-                delete ripTrack;
-        }
-
-        m_artistEdit->SetText(m_artistName);
-        m_albumEdit->SetText(m_albumName);
-        m_genreEdit->SetText(m_genreName);
-        m_yearEdit->SetText(m_year);
-        m_compilationCheck->SetCheckState(isCompilation);
-
-        if (!isCompilation)
-            m_switchTitleArtist->SetVisible(false);
-        else
-            m_switchTitleArtist->SetVisible(true);
-    }
+    if (!m_isCompilation)
+        m_switchTitleArtist->SetVisible(false);
+    else
+        m_switchTitleArtist->SetVisible(true);
 
     BuildFocusList();
     updateTrackList();
@@ -895,12 +848,50 @@ void Ripper::scanCD(void)
     }
 #endif // HAVE_CDIO
 
-    if (m_decoder)
-        delete m_decoder;
-
+    delete m_decoder;
     m_decoder = new CdDecoder("cda", nullptr, nullptr);
-    if (m_decoder)
-        m_decoder->setDevice(m_CDdevice);
+    if (!m_decoder)
+        return;
+
+    m_decoder->setDevice(m_CDdevice);
+
+    m_tracks->clear();
+    for (int trk = 0, trks = m_decoder->getNumTracks(); trk < trks; ++trk)
+    {
+        MusicMetadata *metadata = m_decoder->getMetadata(trk + 1);
+        if (metadata)
+        {
+            RipTrack *ripTrack = new RipTrack;
+            ripTrack->metadata = metadata;
+            ripTrack->length = metadata->Length();
+
+            if (metadata->Compilation())
+            {
+                m_isCompilation = true;
+                m_artistName = metadata->CompilationArtist();
+            }
+            else if (m_artistName.isEmpty())
+            {
+                m_artistName = metadata->Artist();
+            }
+
+            if (m_albumName.isEmpty())
+                m_albumName = metadata->Album();
+
+            if (m_genreName.isEmpty() && !metadata->Genre().isEmpty())
+                m_genreName = metadata->Genre();
+
+            if (m_year.isEmpty() && metadata->Year() > 0)
+                m_year = QString::number(metadata->Year());
+
+            QString title = metadata->Title();
+            ripTrack->isNew = isNewTune(m_artistName, m_albumName, title);
+
+            ripTrack->active = ripTrack->isNew;
+
+            m_tracks->push_back(ripTrack);
+        }
+    }
 }
 
 void Ripper::deleteAllExistingTracks(void)
diff --git a/mythplugins/mythmusic/mythmusic/cdrip.h b/mythplugins/mythmusic/mythmusic/cdrip.h
index f1c573c..b65d5f1 100644
--- a/mythplugins/mythmusic/mythmusic/cdrip.h
+++ b/mythplugins/mythmusic/mythmusic/cdrip.h
@@ -166,6 +166,7 @@ class Ripper : public MythScreenType
     QVector<RipTrack*> *m_tracks;
 
     QString            m_albumName, m_artistName, m_genreName, m_year;
+    bool               m_isCompilation;
     QStringList        m_searchList;
     bool               m_somethingwasripped;
     bool               m_mediaMonitorActive;
-- 
1.7.9.5

